subprojects/core/src/main/java/org/gradle/internal/service/scopes/BuildScopeServices.java

```diff
@@ -416,6 +425,44 @@ protected ITaskFactory createITaskFactory(Instantiator instantiator, TaskClassIn
             new TaskFactory());
     }
 
+    protected GroovyScriptClasspathProvider createGroovyScriptClasspathProvider(
+        ClassLoaderScopeRegistry classLoaderScopeRegistry,
+        DependencyManagementServices dependencyManagementServices,
+        FileResolver fileResolver,
+        FileCollectionFactory fileCollectionFactory,
+        DependencyMetaDataProvider dependencyMetaDataProvider,
+        DependencyFactory dependencyFactory
+    ) {
+        DependencyResolutionServices resolutionServices = dependencyManagementServices.create(
+            fileResolver,
+            fileCollectionFactory,
+            dependencyMetaDataProvider,
+            new UnknownProjectFinder("Some unknown project"),
+            RootScriptDomainObjectContext.PLUGINS
+        );
+        return new GroovyScriptClasspathProvider(
+            classLoaderScopeRegistry.getCoreAndPluginsScope(),
+            () -> {
+                String version = System.getProperty("org.gradle.api.version");
+                return version == null
+                    ? gradleApisFromCurrentGradle(dependencyFactory)
+                    : gradleApisFromRepository(resolutionServices, dependencyFactory, version);
+            },
+            () -> ((SelfResolvingDependency) dependencyFactory.createDependency(DependencyFactory.ClassPathNotation.LOCAL_GROOVY)).resolve());
```

Comment:
Looks like we are using local Groovy regardless of the Gradle version, but that will change. Shouldn't we resolve the Groovy version via the repository instead if the target API version is specified?

Reply:
I wouldn't resolve a different Groovy/Kotlin version per API. If we'd do this, then we'd also need to upgrade the Kotlin/Groovy code and use a different compiler as well. I don't see why we can't compile the old code with the newer Kotlin/Groovy version.

Reply:
I think that reasoning makes sense. Let's mention it here in a comment.